File: 프로그래머스/3/72413. 합승 택시 요금/합승 택시 요금.py
from heapq import heappop, heappush

def solution(n, s, a, b, fares):
    graph = [[] for _ in range(n+1)]
    for u, v, w in fares:
        graph[u].append((v, w))
        graph[v].append((u, w))
        
    distances_s = dijkstra(graph, n, s)
    distances_a = dijkstra(graph, n, a)
    distances_b = dijkstra(graph, n, b)
    
    return min(map(lambda idx: distances_s[idx] + distances_a[idx] + distances_b[idx], range(n+1)))


def dijkstra(graph, n, start):
    INF = float('inf')
    distances = [INF] * (n+1)
    distances[start] = 0
    
    hq = []
    heappush(hq, (0, start))
    
    while hq:
        dist, now = heappop(hq)
        
        if distances[now] < dist:
            continue
            
        for nxt, nd in graph[now]:
            nxt_distance = nd + dist
            
            if distances[nxt] > nxt_distance:
                distances[nxt] = nxt_distance
                heappush(hq, (nxt_distance, nxt))
                
    return distances

File: 프로그래머스/3/72413. 합승 택시 요금/합승 택시 요금.java
import java.util.*;


class Solution {
    public int solution(int n, int s, int a, int b, int[][] fares) {
        List<List<int[]>> graph = new ArrayList<>();
        for (int i = 0; i <= n; i++)
            graph.add(new ArrayList<>());
        
        for (int[] fare: fares) {
            int u = fare[0];
            int v = fare[1];
            int w = fare[2];
            
            graph.get(u).add(new int[]{v, w});
            graph.get(v).add(new int[]{u, w});
        }
        
        int[] distancesS = dijkstra(graph, n, s);
        int[] distancesA = dijkstra(graph, n, a);
        int[] distancesB = dijkstra(graph, n, b);
        
        System.out.println(Arrays.toString(distancesS));
        System.out.println(Arrays.toString(distancesA));
        System.out.println(Arrays.toString(distancesB));
        
        int mnDist = (int)1e8;
        for (int idx = 0; idx <= n; idx++) {
            int nowDist = distancesS[idx] + distancesA[idx] + distancesB[idx];
            mnDist = Math.min(mnDist, nowDist);
        }
        
        return mnDist;
    }
    
    public int[] dijkstra(List<List<int[]>> graph, int n, int start) {
        int INF = (int)1e8;  // 1000000000
        int[] distances = new int[n+1];
        Arrays.fill(distances, INF);  // 배열을 전부 INF 로 초기화
        distances[start] = 0;
        
        PriorityQueue<int[]> pq = new PriorityQueue<>((o1, o2) -> o1[0] - o2[0]);  // (거리, 현재 노드)
        pq.add(new int[]{0, start});
        
        while (!pq.isEmpty()) {
            int[] first = pq.poll();
            
            int distance = first[0];
            int now = first[1];
            
            if (distances[now] < distance)
                continue;
            
            for (int[] nxtNode: graph.get(now)) {
                int nxt = nxtNode[0];
                int nxtDistance = distance + nxtNode[1];
                
                if (distances[nxt] > nxtDistance) {
                    distances[nxt] = nxtDistance;
                    pq.add(new int[]{nxtDistance, nxt});
                }
            } 
        }
        
        return distances;
    }
}